### ISSUE TITLE
Nested groups freezes

### ISSUE THREAD

**Issue by sysradium** [celery/celery#3009]
I use Celery 3.1.19 + RabbitMQ + Redis (result storage) and Python 3.5.1

Basically I have a task which does something and the passes results to another two tasks, which should run in parallel. Something like that:

     some_chain = (get_ip_address.s('192.168.1.1') | group(check1.s(), check2.s())

Now I want these `some_chain` to run n parallel for an ip list. So I put that into group:

    group(get_ip_address.s(ip) | group(check1.s(), check2.s()) for in ip_address_list)().get()

Even when all tasks are complete `get()` which uses `iter_native` never ends.  `join()` does not work either.

In case of `get()` call Keyboard interruption produces this traceback:

```
/usr/local/lib/python3.5/site-packages/celery/result.py in get(self, timeout, propagate, interval, callback, no_ack)
    576         return (self.join_native if self.supports_native_join else self.join)(
    577             timeout=timeout, propagate=propagate,
--> 578             interval=interval, callback=callback, no_ack=no_ack)
    579
    580     def join(self, timeout=None, propagate=True, interval=0.5,

/usr/local/lib/python3.5/site-packages/celery/result.py in join_native(self, timeout, propagate, interval, callback, no_ack)
    686         )
    687         acc = None if callback else [None for _ in range(len(self))]
--> 688         for task_id, meta in self.iter_native(timeout, interval, no_ack):
    689             value = meta['result']
    690             if propagate and meta['status'] in states.PROPAGATE_STATES:

/usr/local/lib/python3.5/site-packages/celery/backends/base.py in get_many(self, task_ids, timeout, interval, no_ack, READY_STATES)
    479             if timeout and iterations * interval >= timeout:
    480                 raise TimeoutError('Operation timed out ({0})'.format(timeout))
--> 481             time.sleep(interval)  # don't busy loop.
    482             iterations += 1
    483

KeyboardInterrupt:
```

With `join()` this:
```
/usr/local/lib/python3.5/site-packages/celery/result.py in join(self, timeout, propagate, interval, callback, no_ack)
    640             value = result.get(
    641                 timeout=remaining, propagate=propagate,
--> 642                 interval=interval, no_ack=no_ack,
    643             )
    644             if callback:

/usr/local/lib/python3.5/site-packages/celery/result.py in get(self, timeout, propagate, interval, no_ack, follow_parents, EXCEPTION_STATES, PROPAGATE_STATES)
    167             interval=interval,
    168             on_interval=on_interval,
--> 169             no_ack=no_ack,
    170         )
    171         if meta:

/usr/local/lib/python3.5/site-packages/celery/backends/base.py in wait_for(self, task_id, timeout, interval, no_ack, on_interval)
    221                 on_interval()
    222             # avoid hammering the CPU checking status.
--> 223             time.sleep(interval)
    224             time_elapsed += interval
    225             if timeout and time_elapsed >= timeout:

KeyboardInterrupt:
```

Why does it think that the task isn't ready? In fact I could avoid this problem not using group:

```
def wait_for_tasks(tasks):
    while True:
        if all(t.ready() for t in tasks):
            print('Complete')
            return [t.get() for t in tasks]

tasks = [(get_ext_ip.s(ip) | group(check1.s(), check2())).delay() for ip in ip_address_list]
results = wait_for_tasks(tasks)
```

But does not look nice. I tried both with eventlet and prefork.

#### Update

Tried to use Celery from master. Chains do not work there at all. 

```
(get_ext_ip.s('192.168.1.1') | check1.s() | check2.s())().get()
```
Will freeze forever. Debug output it console shows that that `check1` and `check2` are not even called.

**Comment by ask** [comment]
Works fine for me in master here, with `redis://` as the result backend:

```
In [2]: (add.s(2, 2) | add.s(4) | add.s(8))().get()
Out[2]: 16
```

Are you sure you are able to store and retrieve results at all?

**Comment by ask** [comment]
Sorry, I reproduced the issue in master when using task_protocol=1.

It's fixed in master now.

You say the issue only manifests in 3.1 when you use a group?

I'm not sure what the outer group is for here, why do you not just use:

    (get_ip_address.s(ip) | group(check1.s(), check2.s()) for in ip_address_list)().get()


Note that any group chained to another task (e.g. `group(...) | task.s()`) is automatically upgraded to a *chord*, so if you use the `rpc://` result backend which does not currently support chords, it will deadlock.

**Comment by syradium** [comment]
@ask thank's for the fix. It has the python 3.5 bugfix with ordered dict but I couldn't use it because of the chain issue.

The intention behind that outer group was to parallelise jobs. So I thought that this:

```
(get_ip_address.s(ip) | group(check1.s(), check2.s()) for in ip_address_list)().get()
```

Will just run this chains one after the other.

> Note that any group chained to another task (e.g. group(...) | task.s()) is automatically upgraded to a chord, so if you use the rpc:// result backend which does not currently support chords, it will deadlock.

Yes, I know that. Does it apply to situations like mine, where the callback isn't just a task - it's a group of tasks or a chain?

> You say the issue only manifests in 3.1 when you use a group?

Compared to master? I made a lot of attempts, do not remember exactly but as far as I remember I could not try that on master because chains did not work. I need to try this out again with the newest one.

**Comment by ask** [comment]
Closing this, as we don't have the resources to complete this task.

May be fixed in master, let's see if comes back after 4.0 release.
